notifications/source/client.go

```diff
@@ -0,0 +1,187 @@
+var ErrRulesOutdated = fmt.Errorf("rules version is outdated")
+
+// BasicAuthTransport is an http.RoundTripper that adds basic authentication and a User-Agent header to HTTP requests.
+type BasicAuthTransport struct {
+	http.RoundTripper // RoundTripper is the underlying HTTP transport to use for making requests.
+
+	Username   string
+	Password   string
+	ClientName string // ClientName is used to set the User-Agent header.
+}
+
+// RoundTrip adds basic authentication headers to the request and executes the HTTP request.
+func (b *BasicAuthTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.SetBasicAuth(b.Username, b.Password)
+	// As long as our round tripper is used for the client, the User-Agent header below
+	// overrides any other value set by the user.
+	req.Header.Set("User-Agent", b.ClientName)
+
+	return b.RoundTripper.RoundTrip(req)
+}
+
+// Client provides a common interface to interact with the Icinga Notifications API.
+// It holds the configuration for the API endpoint and the HTTP client used to make requests.
+type Client struct {
+	cfg Config // cfg holds base API endpoint URL and authentication details.
+
+	client http.Client // HTTP client used for making requests to the Icinga Notifications API.
+
+	IcingaWebBaseUrl *url.URL // IcingaWebBaseUrl holds the base URL for Icinga Web 2.
+	Endpoints        struct {
+		ProcessEvent string // ProcessEvent holds the URL for the process event endpoint.
+	}
+}
+
+// NewClient creates a new Client instance with the provided configuration.
+//
+// The projectName is used to set the User-Agent header in HTTP requests sent by this client and should be
+// set to the name of the project using this client (e.g., "Icinga DB").
+//
+// It may return an error if the API base URL or Icinga Web 2 base URL cannot be parsed.
+func NewClient(cfg Config, projectName string) (*Client, error) {
+	client := &Client{
+		cfg: cfg,
+		client: http.Client{
+			//Timeout: cfg.Timeout, // Uncomment once Timeout is (should be?) user configurable.
+			Transport: &BasicAuthTransport{
+				RoundTripper: http.DefaultTransport,
+				Username:     cfg.Username,
+				Password:     cfg.Password,
+				ClientName:   projectName,
+			},
+		},
+	}
+
+	baseUrl, err := url.Parse(cfg.ApiBaseUrl)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to parse API base URL")
+	}
+
+	client.Endpoints.ProcessEvent = baseUrl.ResolveReference(&url.URL{Path: "/process-event"}).String()
+
+	client.IcingaWebBaseUrl, err = url.Parse(cfg.IcingaWeb2BaseUrl)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to parse Icinga Web 2 base URL")
+	}
+
+	return client, nil
+}
+
+// ProcessEvent submits an event to the Icinga Notifications /process-event API endpoint.
+//
+// It serializes the event into JSON and sends it as a POST request to the process event endpoint. In most cases, the
+// Event.RulesVersion and Event.RuleIds must be set.
```

Comment:
> In most cases

This isn't very helpful if the whole doc comment doesn't mention the rare other case.

Isn't the other case the first event sent by a source, so the source will typically even need to do that, so that should be mentioned.

Reply:
Reworded.